manila/share/drivers/nexenta/ns5/nexenta_nas.py

```diff
@@ -442,34 +442,45 @@ def update_access(self, context, share, access_rules, add_rules,
             self._update_cifs_access(share, add_rules, delete_rules)
 
     def _update_nfs_access(self, share, rw_list, ro_list):
-        security_contexts = []
+        # Define allowed security context types to be able to tell whether
+        # the 'security_contexts' dict contains any rules at all
+        context_types = {'none', 'root', 'readOnlyList', 'readWriteList'}
+
+        security_contexts = {'securityModes': ['sys']}
+
+        def add_sc(addr_list, sc_type):
+            if sc_type not in context_types:
+                return
+
+            rule_list = []
 
-        def append_sc(addr_list, sc_type):
             for addr in addr_list:
                 address_mask = addr.strip().split('/', 1)
                 address = address_mask[0]
-                ls = [{"allow": True, "etype": "fqdn", "entity": address}]
+                ls = {"allow": True, "etype": "fqdn", "entity": address}
                 if len(address_mask) == 2:
                     try:
                         mask = int(address_mask[1])
                         if 0 <= mask < 31:
-                            ls[0]['mask'] = mask
-                            ls[0]['etype'] = 'network'
+                            ls['mask'] = mask
+                            ls['etype'] = 'network'
                     except Exception:
                         raise exception.InvalidInput(
                             reason=_(
                                 '<{}> is not a valid access parameter').format(
                                     addr))
-                new_sc = {"securityModes": ["sys"]}
-                new_sc[sc_type] = ls
-                security_contexts.append(new_sc)
+                rule_list.append(ls)
+
+            # Prevent setting a context type without any addresses, this results in an API error
+            if rule_list:
+                security_contexts[sc_type] = rule_list
 
-        append_sc(rw_list, 'readWriteList')
-        append_sc(ro_list, 'readOnlyList')
-        payload = {"securityContexts": security_contexts}
+        add_sc(rw_list, 'readWriteList')
+        add_sc(ro_list, 'readOnlyList')
+        payload = {'securityContexts': [security_contexts]}
         share_path = self._get_dataset_path(share)
         if self.nef.nfs.list({'filesystem': share_path}):
-            if not security_contexts:
+            if not set(security_contexts.keys()) & context_types:
```

Comment:
what is the point of ```& context_types``` here?

Reply:
I'm doing an intersect between two sets - the keys in `security_contexts` dict and the above defined `context_types` set - to see whether we have added any security context rules. If we have not, then the initial `ro_list` and `rw_list` was probably empty or the rules were not valid.

You see, we can no longer check `if not security_contexts` expression, because we unconditionally define the dict at the beginning of the function with a `securityModes` key. That's why I'm checking for any additional keys from the `context_types` set, revealing whether we had added at least one access rule.

Reply:
Correct me if I am missing something, but context_types will always be non-empty dict since it is pre-defined in the beginning of the function, which makes 
```
if not set(security_contexts.keys()) & context_types
```
equal to 
```
if not set(security_contexts.keys())
```